kolibri/core/content/utils/import_export_content.py

```diff
@@ -155,52 +200,157 @@ def get_import_export_data(  # noqa: C901
-                )
-
-                file_objects = LocalFile.objects.filter(
-                    files__contentnode__in=segment_topics,
-                ).values("id", "file_size", "extension")
-                if available is not None:
-                    file_objects = file_objects.filter(available=available)
-                for f in file_objects:
-                    queried_file_objects[f["id"]] = f
-
-        min_boundary += dynamic_chunksize
-
     files_to_download = list(queried_file_objects.values())
 
     total_bytes_to_transfer = sum(map(lambda x: x["file_size"] or 0, files_to_download))
     return number_of_resources, files_to_download, total_bytes_to_transfer
 
 
+def get_content_nodes_selectors(channel_id, nodes_queries_list):
+    """
+    Returns a dictionary with a set of include_node_ids and exclude_node_ids
+    that can be used with the given channel_id to import the nodes contained
+    in nodes_queries_list.
+    """
+
+    include_node_ids = list()
+    exclude_node_ids = list()
+
+    channel_metadata = ChannelMetadata.objects.get(id=channel_id)
+
+    available_node_ids = set(
+        itertools.chain.from_iterable(
+            nodes_query.values_list("id", flat=True)
+            for nodes_query in nodes_queries_list
+        )
+    )
+
+    available_nodes_queue = [channel_metadata.root]
+
+    while len(available_nodes_queue) > 0:
+        node = available_nodes_queue.pop(0)
+
+        # We could add nodes to exclude_node_ids when less than half of the
+        # sibling nodes are missing. However, it is unclear if this would
+        # be useful.
+
+        if node.kind == "topic":
+            leaf_node_ids = _get_leaf_node_ids(node)
+            matching_leaf_nodes = leaf_node_ids.intersection(available_node_ids)
+            missing_leaf_nodes = leaf_node_ids.difference(available_node_ids)
+            if len(missing_leaf_nodes) == 0:
+                assert node.id not in include_node_ids
+                include_node_ids.append(node.id)
+            elif len(matching_leaf_nodes) > 0:
+                available_nodes_queue.extend(node.children.all())
+        elif node.id in available_node_ids:
+            assert node.id not in include_node_ids
+            include_node_ids.append(node.id)
+
+    return {
+        "id": channel_id,
+        "version": channel_metadata.version,
+        "include_node_ids": include_node_ids,
+        "exclude_node_ids": exclude_node_ids,
+    }
+
+
+def update_content_manifest(manifest_file, nodes_selectors):
+    try:
+        with open(manifest_file, "r") as fp:
+            manifest_data = json.load(fp)
+    except (FileNotFoundError, ValueError):
+        # Use ValueError rather than JSONDecodeError for Py2 compatibility
+        manifest_data = None
+
+    if not isinstance(manifest_data, dict):
+        manifest_data = {}
+
+    channels = manifest_data.setdefault("channels", [])
+
+    # TODO: If the channel is already listed, it would be nice to merge the
```

Comment:
So, in the case the channel is already in the manifest what happens, do we have two entries for the same channel in the manifest?

Reply:
Yeah, exactly. So it could end up with a manifest file like this:

```
{
    "channels": [
        {
            "id": "e409b964366a59219c148f2aaa741f43",
            "version": 10,
            "include_node_ids": [
                "6277aa0c44235435acdc8a9ed98f466b"
            ],
            "exclude_node_ids": []
        },
        {
            "id": "e409b964366a59219c148f2aaa741f43",
            "version": 10,
            "include_node_ids": [
                "e409b964366a59219c148f2aaa741f43"
            ],
            "exclude_node_ids": []
        }
    ],
    "channel_list_hash": "a2c59ee9bab4586a78e56ca8cf13a912"
}
```

Where the first time around, I exported just "6277aa0c44235435acdc8a9ed98f466b", and the second time I exported the entire channel.

Note that identical entries are skipped, but if we have a different selection for the same channel, they each get a new entry. It would be better if it merges them more exhaustively. For instance the second entry is a superset of the first entry. So, I added a TODO about that. Could do that by adding the existing entries to our list of selected nodes.

Come to think of it, that doesn't sound too difficult to implement, so I think I'll give it a shot right now :)

Reply:
Yes - I was wondering if we could just do the TODO now, and not leave it for later!

Reply:
I think so, indeed. Ran out of time today, but I'll let you know how it goes tomorrow!

Reply:
Okay, I added a commit with a couple things:

- Removed `exclude_node_ids` from manifest.json. It isn't being used right now, and it makes our life easier if it isn't there. If we are only _including_ nodes, we can trivially extend that set of nodes when a new export happens. Otherwise (if there is the possibility of nodes being excluded), we would have to expand the selection into a complete set of nodes and then simplify it again.
- Added a `content_manifest` module to deal with this in one place, moving some file access and json parsing code out of import_export_content.py. It is responsible for reading and writing manifest files, as well as simplifying the provided set of node IDs.